src/Fields/set!.jl

```diff
@@ -39,26 +39,14 @@ function set!(u::Field, f::Union{Array, CuArray, OffsetArray})
 end
 
 function set!(u::Field, v::Field)
+    # Note: we only copy interior points.
+    # To copy halos use `parent(u) .= parent(v)`.
+    
     if architecture(u) === architecture(v)
-        try # to transfer halos between architectures
-            parent(u) .= parent(v)
-        catch # just copy interior points
-            interior(u) .= interior(v)
-        end
+        interior(u) .= interior(v)
     else
-        try # to transfer halos between architectures
-            u_parent = parent(u)
-            v_parent = parent(v)
-            # If u_parent is a view, we have to convert to an Array.
-            # If u_parent or v_parent is on the GPU, we don't expect
-            # SubArray.
-            u_parent isa SubArray && (u_parent = Array(u_parent))
-            v_parent isa SubArray && (v_parent = Array(v_parent))
-            copyto!(u_parent, v_parent)
-        catch # just copy interior points
-            v_data = arch_array(architecture(u), v.data)
-            interior(u) .= interior(v_data, location(v), v.grid, v.indices)
-        end
+        v_data = arch_array(architecture(u), v.data)
+        interior(u) .= interior(v_data, location(v), v.grid, v.indices)
```

Comment:
so only interior points are copied, right?
then 

https://github.com/CliMA/Oceananigans.jl/blob/edb4268d34e86ec9e202649307e53444ef92c5e7/test/test_nonhydrostatic_models.jl#L150-L154

will fail!

Do we want `set!` to call `fill_halos!`?

Reply:
Why not simply:

```julia
set!(u::Field, v::Field) = set!(u::Field, v.data)
```

Reply:
what `set!` does that call? Maybe that will work!

Reply:
I guess this:
https://github.com/CliMA/Oceananigans.jl/blob/87a046fdd7da4358e6da2b1774ddc7c258f87948/src/Fields/set!.jl#L35-L39

Reply:
That should work

Reply:
https://github.com/CliMA/Oceananigans.jl/blob/3557db3692ff8e1978b126369230c9bacae42ae8/src/Fields/broadcasting_abstract_fields.jl#L26-L39

Reply:
I don't think `Base.Broadcast.materialize!(dest.data, bc)` does what we want; it includes halo regions. So I think we need to keep broadcasting over `interior` when setting fields.

Reply:
Side note I think we need to clean up broadcasting...

Reply:
Yes I had troubles with `set!(u::Field, v::Field) = set!(u::Field, v.data)` over in ParameterEstimocean.jl. Let's leave it as is?

Reply:
But what about the halos? They are not copied like this. Should we change the tests?

https://github.com/CliMA/Oceananigans.jl/blob/edb4268d34e86ec9e202649307e53444ef92c5e7/test/test_nonhydrostatic_models.jl#L150-L154